Support to SPM in nef-playground

miguelangel-dev opened the pull request:
## Issues
Related: #33 

## Details
This PR adds support to Swift Packages in **playground component**. It enables a new flag in `nef playground` to use swift-package as a dependency manager.

```bash
--spm                   Use Swift Packages as a dependency manager.
```

Example of use:
```
nef playground --name Chiro --spm
```

![Screen Shot 2020-07-31 at 15 03 10](https://user-images.githubusercontent.com/25568562/89037563-02b46a80-d33f-11ea-8ea4-d229dbd42252.png)

miguelangel-dev committed:
update project to Xcode 12

miguelangel-dev committed:
Swift packages will be compiled together with workspace

minor changes

miguelangel-dev committed:
updated swift-argument-parser dependency

miguelangel-dev committed:
add templates for SPM - osx version

miguelangel-dev committed:
update nef Playgrounds module to support SPM option

miguelangel-dev committed:
update project manipulator (scripts) to support SPM template

miguelangel-dev committed:
update swift-package osx template

minor changes

miguelangel-dev committed:
fix extract scheme: it contains white-spaces

miguelangel-dev committed:
set macos version

miguelangel-dev committed:
fix nef version

miguelangel-dev committed:
update clean templates to support SPM

miguelangel-dev committed:
update Package project - update swift-argument-parser

miguelangel-dev committed:
add templates for SPM - ios version

miguelangel-dev committed:
update SPM templates to support Xcode12

update macos sdk version in nef playground templates

minor changes

miguelangel-dev committed:
fix typo

miguelangel-dev committed:
set Xcode playground dependencies default to SPM

miguelangel-dev committed:
fix merge

truizlop commented on project/Component/NefCompiler/Instances/NefCompilerSystem.swift:
```diff
@@ -64,7 +64,6 @@ class NefCompilerSystem: CompilerSystem {
         binding(
             |<-self.buildPods(xcworkspace: xcworkspace, platform: platform, cached: cached),
             |<-self.buildCarthage(xcworkspace: xcworkspace, platform: platform, cached: cached),
-            |<-self.buildSPM(xcworkspace: xcworkspace, platform: platform, cached: cached),
```
Why is this not build anymore?

miguelangel-dev replied:
The reason is that Swift Package dependencies will be resolved during the xcode-build step, so we do not need to compile the external dependencies -in a different step- as we need to CocoaPods or Carthage.

truizlop commented on project/UI/Carbon/Carbon.swift:
```diff
@@ -22,23 +22,23 @@ public struct CarbonCommand: ParsableCommand {
     
-    @ArgumentParser.Option(name: .customLong("show-lines"), default: true, help: "Shows/hides lines of code [true | false]")
-    private var lines: Bool
+    @ArgumentParser.Option(name: .customLong("show-lines"), help: "Shows/hides lines of code [true | false]")
```
This should be a flag with the corresponding inversion property.

miguelangel-dev replied:
mmmm it is weird, I thought it has done! I will review it

miguelangel-dev replied:
Yep, it was done in a reviewed [PR](https://github.com/bow-swift/nef/pull/167#discussion_r463543065) about command-line autocomplete. It was a previous one - you can see the pending-to-merge branch `shell_autocomplete+spm_nefc` where it was merged.

truizlop commented on project/UI/Carbon/Carbon.swift:
```diff
@@ -22,23 +22,23 @@ public struct CarbonCommand: ParsableCommand {
     
-    @ArgumentParser.Option(name: .customLong("show-watermark"), default: true, help: "Shows/hides the watermark [true | false]")
-    private var watermark: Bool
+    @ArgumentParser.Option(name: .customLong("show-watermark"), help: "Shows/hides the watermark [true | false]")
```
This should also be a flag with the corresponding inversion property.

miguelangel-dev replied:
it was done in a previous PR

truizlop commented on project/UI/CarbonPage/CarbonPage.swift:
```diff
@@ -22,26 +22,26 @@ public struct CarbonPageCommand: ParsableCommand {
     
-    @ArgumentParser.Option(name: .customLong("show-lines"), default: true, help: "Shows/hides lines of code [true | false]")
-    private var lines: Bool
+    @ArgumentParser.Option(name: .customLong("show-lines"), help: "Shows/hides lines of code [true | false]")
```
This should be a flag.

miguelangel-dev replied:
it was done in a previous PR

truizlop commented on project/UI/CarbonPage/CarbonPage.swift:
```diff
@@ -22,26 +22,26 @@ public struct CarbonPageCommand: ParsableCommand {
     
-    @ArgumentParser.Option(name: .customLong("show-watermark"), default: true, help: "Shows/hides the watermark [true | false]")
-    private var watermark: Bool
+    @ArgumentParser.Option(name: .customLong("show-watermark"), help: "Shows/hides the watermark [true | false]")
```
This should be a flag.

miguelangel-dev replied:
it was done in a previous PR

truizlop commented on project/UI/Carbon/Carbon.swift:
```diff
@@ -22,23 +22,23 @@ public struct CarbonCommand: ParsableCommand {
     
-    @ArgumentParser.Option(default: "nef", help: "Background color in hexadecimal")
-    private var background: String
+    @ArgumentParser.Option(help: "Background color in hexadecimal")
```
This help message is a bit misleading, as you can also use named colors.

truizlop replied:
```suggestion
    @ArgumentParser.Option(help: "Background color (hexadecimal or predefined)")
```

truizlop commented on project/UI/CarbonPage/CarbonPage.swift:
```diff
@@ -22,26 +22,26 @@ public struct CarbonPageCommand: ParsableCommand {
     
-    @ArgumentParser.Option(default: "nef", help: "Background color in hexadecimal")
-    private var background: String
+    @ArgumentParser.Option(help: "Background color in hexadecimal")
```
This help message is a bit misleading, as you can also use named colors.

miguelangel-dev replied:
With auto-completion it will suggest you use one of the predefined colors, but do you have any suggestion to message?  - I was thinking about to a simple `Background color` but I think it is better to add the expected format, so maybe we could say `Background color in hexadecimal or predefined` wdut?

truizlop commented on project/UI/Jekyll/Jekyll.swift:
```diff
@@ -21,8 +21,8 @@ public struct JekyllCommand: ParsableCommand {
-    @ArgumentParser.Option(name: .customLong("main-page"), default: "README.md", help: "Path to 'README.md' file to be used as the index page")
-    private var mainPage: String
+    @ArgumentParser.Option(name: .customLong("main-page"), help: "Path to 'README.md' file to be used as the index page")
+    private var mainPage: String = "README.md"
```
I don't think this should have a default value. If this option is not provided and the file `README.md` does not exist, what is the expected outcome?

miguelangel-dev replied:
As Jekyll works using our sidebar template always needs a `main-page` - in case you set your main page it will take it, in another case it will create a new one called `README.md` with empty content.

truizlop commented on template/ios/carthage/PROJECT.playground/Pages/page.xcplaygroundpage/Contents.swift:
```diff
@@ -1 +1 @@
-import Foundation
+import SwiftUI
```
Why do you need to import SwiftUI here? It creates an unnecessary dependency.

miguelangel-dev replied:
Yep we can remove the whole dependencies...

truizlop commented on template/ios/carthage/PROJECT.playground/Sources/NefPlaygroundSupport.swift:
```diff
@@ -1,4 +1,4 @@
-import UIKit
+import SwiftUI
```
Why do you need to import SwiftUI here? It creates an unnecessary dependency.

miguelangel-dev replied:
In fact, it is not necessary, we can continue using UIKit

miguelangel-dev committed:
remove unuseful dependencies from templates

Update project/UI/Carbon/Carbon.swift

Co-authored-by: Tomás Ruiz-López <truizlop@gmail.com>

Update project/UI/Playground/Playground.swift

Co-authored-by: Tomás Ruiz-López <truizlop@gmail.com>

Update project/UI/Playground/Playground.swift

Co-authored-by: Tomás Ruiz-López <truizlop@gmail.com>

Update project/UI/CarbonPage/CarbonPage.swift

Co-authored-by: Tomás Ruiz-López <truizlop@gmail.com>

truizlop commented on project/UI/CarbonPage/CarbonPage.swift:
```diff
@@ -22,26 +22,26 @@ public struct CarbonPageCommand: ParsableCommand {
     
-    @ArgumentParser.Option(default: "nef", help: "Background color in hexadecimal")
-    private var background: String
+    @ArgumentParser.Option(help: "Background color in hexadecimal")
```
```suggestion
    @ArgumentParser.Option(help: "Background color (hexadecimal or predefined)")
```

truizlop commented on project/UI/Playground/Playground.swift:
```diff
@@ -15,18 +15,21 @@ public struct PlaygroundCommand: ParsableCommand {
 
-    @ArgumentParser.Option(default: .ios, help: "set the target to `ios` or `macos`")
-    private var platform: Platform
+    @ArgumentParser.Option(help: "set the target to `ios` or `macos`")
```
```suggestion
    @ArgumentParser.Option(help: "Set the target to `ios` or `macos`")
```

truizlop commented on project/UI/Playground/Playground.swift:
```diff
@@ -15,18 +15,21 @@ public struct PlaygroundCommand: ParsableCommand {
     @ArgumentParser.Option(help: ArgumentHelp("Xcode Playground to be transformed into nef Playground", valueName: "Xcode Playground"))
     private var playground: ArgumentPath?
     
+    @ArgumentParser.Flag(help: "Use Swift Packages as a dependency manager")
```
```suggestion
    @ArgumentParser.Flag(help: "Use Swift Package Manager to resolve dependencies")
```

truizlop commented on project/UI/Playground/Playground.swift:
```diff
@@ -54,35 +58,34 @@ public struct PlaygroundCommand: ParsableCommand {
         } else if let branch = bowBranch {
-            dependencies = .bow(.branch(branch))
+            return .bow(.branch(branch))
         } else if let commit = bowCommit {
-            dependencies = .bow(.commit(commit))
+            return .bow(.commit(commit))
         } else if let podfileURL = podfile?.url {
-            dependencies = .podfile(podfileURL)
+            return .podfile(podfileURL)
         } else if let cartfileURL = cartfile?.url {
-            dependencies = .cartfile(cartfileURL)
+            return .cartfile(cartfileURL)
+        } else if spm {
+            return .spm
         } else {
-            dependencies = .bow(.version(""))
+            return nil
```
I think this modeling needs to be revisited. This allows an ambiguous input from the user, where you can supply both a Podfile and a Cartfile, and only knowing what is the priority resolving them explains the observed outcome. Also, what happens if I want to use a specific Bow branch, but with SPM?

miguelangel-dev replied:
Right now, three dependency managers are mutually exclusive. If you set more than one, the behaviour will be not expected because it is not supported; so the order at this point is not important. Following your doubt, I have just thought we could detect that invalid scenario and throw a `fatal-error` (instead of picking one randomly, as we do now)

miguelangel-dev replied:
done!

miguelangel-dev committed:
model invalid options for dependency manager in nef-playground

miguelangel-dev committed:
fix CI: update Xcode version to 12.0.1

minor changes
